# MED5018_Group1
## Project Overview
In this comprehensive study, we have developed a method for analyzing gene sequences with the objective of revealing the characteristics of codon usage patterns across six different species. This method encompasses data cleaning, the calculation of frequencies of codons and their corresponding amino acids, and the visualization of this information. We specifically focused on the preferences in codon usage, a non-random phenomenon observed across species. Through this analysis, we identified both the consistencies and uniqueness in codon preferences among different species, which play a crucial role in various biological processes such as gene expression, cellular functions, RNA processing, protein translation, and folding. This approach offers valuable new insights into the field of molecular genetics, contributing to a deeper understanding of the underlying mechanisms in biological processes.

## Dependencies
* pandas
* numpy
* matplotlib.pyplot
* seaborn
* math
* sklearn.preprocessing.LabelEncoder
* sklearn.cluster.KMeans
* sklearn.decomposition.PCA

## Key Findings
The project yielded several important findings:
##### 1.
##### 2.
##### 3.
## Visualizations 
![mahua](mahua-logo.jpg)


## Team

```javascript
Hanchun.L
```
